### ISSUE TITLE
Admission Webhooks Java DevEx

### ISSUE THREAD

**Issue by csviri** [snowdrop/team#701]
## User Story
User story: As a developer I would like to create and deploy and test an admission web-hook easily as possible. So I don't spend to much time on other aspects then my business logic inside the web-hook.

## Details

Although [Admission controllers](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) are basically HTTP Endpoints, it could be quite time consuming to setup one from scratch from java. 
Started to prototype a microframework that would make it very easy to implement a mutation or a validation hook:
https://github.com/java-operator-sdk/admission-controller-framework

See the definition of a mutation and a validation service, that is then just exposed by an endpoint:

https://github.com/java-operator-sdk/admission-controller-framework/blob/92ebf64de9e5dede7bd5954c7eb165aa08398ab5/samples/quarkus/src/main/java/io/javaoperatorsdk/admissioncontroller/sample/quarkus/AdmissionControllerConfig.java#L16-L33

```java
  @Singleton
  @Named("mutatingController")
  public AdmissionController<Pod> mutatingController() {
    return new AdmissionController<>((resource, operation) -> {
      resource.getMetadata().getLabels().putIfAbsent(APP_NAME_LABEL_KEY, "mutation-test");
      return resource;
    });
  }


  @Singleton
  @Named("validatingController")
  public AdmissionController<Pod> validatingController() {
    return new AdmissionController<>((resource, operation) -> {
      if (resource.getMetadata().getLabels().get(APP_NAME_LABEL_KEY) == null) {
        throw new NotAllowedException("Missing label: " + APP_NAME_LABEL_KEY);
      }
    });
  }
```

(The validation part could be also improved, to have even nicer api)

An other side of the story of Admission Hook  is the Configuration:
```
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  name: "pod-policy.example.com"
webhooks:
- name: "pod-policy.example.com"
  rules:
  - apiGroups:   [""]
    apiVersions: ["v1"]
    operations:  ["CREATE"]
    resources:   ["pods"]
    scope:       "Namespaced"
  clientConfig:
    service:
      namespace: "example-namespace"
      name: "example-service"
    caBundle: "Ci0tLS0tQk...<`caBundle` is a PEM encoded CA bundle which will be used to validate the webhook's server certificate.>...tLS0K"
  admissionReviewVersions: ["v1", "v1beta1"]
  sideEffects: None
  timeoutSeconds: 5
```
The generation of this configuration could be a nice task for **decorate**, like annoting the endpoints and generating the resources with dekorate (see `@MutatingWebhook`,  `@ValidatingWebhook` annotations):

```
  @MutatingWebhook
  @POST
  @Path("mutate")
  @Consumes(MediaType.APPLICATION_JSON)
  @Produces(MediaType.APPLICATION_JSON)
  public AdmissionReview mutate(AdmissionReview admissionReview) {
    return mutationController.handle(admissionReview);
  }

  @ValidatingWebhook(useCertManager=true, operations={CREATE, UPDATE})
  @POST
  @Path("validate")
  @Consumes(MediaType.APPLICATION_JSON)
  @Produces(MediaType.APPLICATION_JSON)
  public AdmissionReview hello(AdmissionReview admissionReview) {
    return validationController.handle(admissionReview);
  }
```

One interesting part would be to generate the required certificate, but also support cert-manager as an alternative. But other several options for configurations are possible.

## Summary

So in summary, to get this work, and able to bootstrap a project, implement a webhook, and deploy it to Kubernetes in no time we need:

1. A Framework that encapsulates the common patters of admission and validation webhooks. Note that this a very thin layer (actually was hesitating if something like this is needed, but for now looks nice, and simple)
2. A generator that generates the project skeleton (maven archetype, cli  or other? ) - this part missing
3. Generate Related Kubernetes Resources Using Dekorate (or Quarkus extension)
  3.1 Deployment, Service 
  3.2 WebhookConfiguration

**Comment by csviri** [comment]
A nice addition to the story would be also support conversion hooks for CRDs.

**Comment by cmoulliard** [comment]
As this proposition needs to be review with Dekorate, ... engineers can you also invite them to discuss this proposition ? @csviri

**Comment by csviri** [comment]
Sure!
cc @iocanel @aureamunoz @Sgitario , please let me know what do you think. Happy to have a call if that needed. 
(This is a cross project issue initiative, that is why I've put it here.)

**Comment by cmoulliard** [comment]
As the proposition concerns different aspects, can you convert it into an EPIC:
- Have a dekorate module generating the `****Webhook`
- Generate the certificate signed by k8s cluster  
- ....

**Comment by csviri** [comment]
Yep, edit the description and add add list the bulletpoints.

**Comment by metacosm** [comment]
The Quarkus extension could take care of generating the config as needed. Of course, a dekorate-based solution would be more generic.

**Comment by csviri** [comment]
Wrote down some bullet points at the end of description. Formulated as a simple user story at the beginning. Can break it down to smaller issues inside the mentioned projects.

(Thx for the feedback looking forward on any other comments, I kinda like to do such brainstorming before jump on further actions )

**Comment by iocanel** [comment]
> Sure! cc @iocanel @aureamunoz @Sgitario , please let me know what do you think. Happy to have a call if that needed. (This is a cross project issue initiative, that is why I've put it here.)

Admission webhooks is something that I haven't played with. I don't even know what they do or what the potential use cases are. So, I'll have to do some reading first and then we could also have a cool.

**Comment by Sgitario** [comment]
> > Sure! cc @iocanel @aureamunoz @Sgitario , please let me know what do you think. Happy to have a call if that needed. (This is a cross project issue initiative, that is why I've put it here.)
> 
> Admission webhooks is something that I haven't played with. I don't even know what they do or what the potential use cases are. So, I'll have to do some reading first and then we could also have a cool.

Same with me

**Comment by aureamunoz** [comment]
> > > Sure! cc @iocanel @aureamunoz @Sgitario , please let me know what do you think. Happy to have a call if that needed. (This is a cross project issue initiative, that is why I've put it here.)
> > 
> > 
> > Admission webhooks is something that I haven't played with. I don't even know what they do or what the potential use cases are. So, I'll have to do some reading first and then we could also have a cool.
> 
> Same with me

Same for me!! 
The project sounds interesting but need to investigate a little bit
